macrosynergy/download/dq_api.py

```diff
@@ -0,0 +1,749 @@
+        :param <dict> proxy: proxy settings for request.
+        """
+        js, success, self.last_url, msg = dq_request(
+            url=url,
+            cert=(self.crt, self.key),
+            headers=self.headers,
+            params=params,
+            proxies=proxy,
+        )
+        self.last_response = {"json": js, "success": success, "msg": msg}
+        return js, success, msg
+
+
+class OAuth(object):
+    """Accessing DataQuery via OAuth.
+
+    :param <str> client_id: string with client id, username.
+    :param <str> client_secret: string with client secret, password.
+    :param <str> url:
+    :param <str> token_url:
+    :param <str> dq_resource_id:
+    """
+
+    def __init__(
+        self,
+        client_id: str,
+        client_secret: str,
+        url: str = OAUTH_BASE_URL,
+        token_url: str = OAUTH_TOKEN_URL,
+        dq_resource_id: str = OAUTH_DQ_RESOURCE_ID,
+        token_proxy: Optional[dict] = None,
+    ):
+
+        self.base_url: str = url
+        self.__token_url: str = token_url
+        self.__dq_api_resource_id: str = dq_resource_id
+
+        id_error = f"client_id argument must be a <str> and not <{type(client_id)}>."
+        assert isinstance(client_id, str), id_error
+        self.client_id: str = client_id
+
+        secret_error = f"client_secret must be a str and not <{type(client_secret)}>."
+        assert isinstance(client_secret, str), secret_error
+
+        self.client_secret: str = client_secret
+        self._stored_token: Optional[dict] = None
+        self.token_data = {
+            "grant_type": "client_credentials",
+            "client_id": self.client_id,
+            "client_secret": self.client_secret,
+            "aud": self.__dq_api_resource_id,
+        }
+
+        # For debugging purposes save last request response.
+        self.status_code: Optional[int] = None
+        self.last_response: Optional[str] = None
+        self.last_url: Optional[str] = None
+        self.token_proxy: Optional[dict] = token_proxy
+
+    def _active_token(self) -> bool:
+        """Confirms if the token being used has not expired."""
+        created: datetime = self._stored_token["created_at"]
+        expires: int = self._stored_token["expires_in"]
+
+        return (datetime.now() - created).total_seconds() / 60 >= (expires - 1)
+
+    def _valid_token(self) -> bool:
+        """Confirms if the credentials passed correspond to a valid token."""
+        return not (self._stored_token is None or self._active_token())
+
+    def _get_token(self) -> str:
+        """Retrieves the token which is used to access DataQuery via OAuth method."""
+        i = 0
+        success = False
+        while (not success) and (i < 3):
+            js, success, self.last_url, msg = dq_request(
+                url=self.__token_url,
+                data=self.token_data,
+                method="post",
+                proxies=self.token_proxy,
+            )
+            i += 1
```

Comment:
Why are we trying to request multiple times?

Reply:
I wonder if we are better off using off the shelves libraries for the OAuth handling:

```python
from oauthlib.oauth2 import BackendApplicationClient
from requests_oauthlib import OAuth2Session

OAUTH_BASE_URL: str = "https://api-developer.jpmorgan.com/research/dataquery-authe/api/v2/"
OAUTH_TOKEN_URL: str = "https://authe.jpmchase.com/as/token.oauth2"
OAUTH_DQ_RESOURCE_ID: str = "JPMC:URI:RS-06785-DataQueryExternalApi-PROD"

class DQOAuth(object):
    def __init__(self, client_id: str, client_secret: str):
         self.client_id: str = client_id
         self.client_secret = client_secret

         self.client = BackendApplicationClient(client_id=client_id, aud=OAUTH_DQ_RESOURCE_ID)
         self.session = OAuth2Session(client=client)

    def _renew_token(self):
         self.session.fetch_token(
             token_url=OAUTH_TOKEN_URL,
             client_id="<client_id>",
            client_secret="<client_secret>"
          )

   def authorized(self):
        return self.session.authorized

    def request(url: str, method: = "GET"):
        if not self.authorized():
             self._renew_token()


        with self.session.request(url=url, method=method) as r:
             assert r.ok
             js = r.json()
        return js
```

Reply:
The logic behind the multiple requests was to retry getting the token 3 times if and when it a token renew request fails. I've reverted to the original method (currently in use).

Reply:
I'll look into using the library methods directly. Seems cleaner to use as well.